Source/Plugin.BLE.iOS/Adapter.cs

```diff
@@ -366,5 +365,141 @@ public static List<AdvertisementRecord> ParseAdvertismentData(NSDictionary adver
 
             return records;
         }
+
+        /// <summary>
+        /// See: https://developer.apple.com/library/archive/documentation/NetworkingInternetWeb/Conceptual/CoreBluetooth_concepts/Art/ReconnectingToAPeripheral_2x.png for a chart of the flow.
+        /// </summary>
+        protected override async Task<IDevice> ConnectNativeAsync(string friendlyName, Guid uuid, Func<IDevice, bool> deviceFilter, CancellationToken cancellationToken = default)
+        {
+            if (uuid == Guid.Empty)
+            {
+                // If we do not have an uuid scan and connect
+                return await ScanAndConnect(friendlyName, deviceFilter);
+            }
+            else
+            {
+                //FYI attempted to use tobyte array insetead of string but there was a problem with byte ordering Guid->NSUuid
+                var nsuuid = new NSUuid(uuid.ToString());
+
+                // If we have an uuid, check if the system can find the device.
+                var peripheral = TryToRetrieveKnownPeripheral(nsuuid);
+                if (peripheral == null)
+                {
+                    // The device haven't been found. We'll try to scan and connect.
+                    return await ScanAndConnect(friendlyName, deviceFilter);
+                }
+                else
+                {
+                    // Try to connect to the found peripheral
+                    var device = await TryToConnectAsync(peripheral);
+                    if (device == null)
+                    {
+                        // Well, it failed, so we'll try to scan again and see if that can repair
+                        return await ScanAndConnect(friendlyName, deviceFilter);
+                    }
+                    else
+                    {
+                        return device;
+                    }
+                }
+            }
+        }
+
+        private async Task<IDevice> ScanAndConnect(string friendlyName, Func<IDevice, bool> deviceFilter)
+        {
+            var peripheral = await ScanForPeripheralAsync(friendlyName, deviceFilter);
+            return await TryToConnectAsync(peripheral);
+        }
+
+        private async Task<CBPeripheral> ScanForPeripheralAsync(string friendlyName, Func<IDevice, bool> deviceFilter)
+        {
+            var taskCompletionSource = new TaskCompletionSource<CBPeripheral>();
+            var stopToken = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+            EventHandler<DeviceEventArgs> handler = (sender, args) =>
+            {
+                var peripheral = args.Device.NativeDevice as CBPeripheral;
```

Comment:
same as earlier -> own method?

Reply:
I guess this could be unhooked right after taskCompletionSource.TrySetResult(args.Device); is called, if the handler is declared before the lambda expression